pandas/_libs/index.pyx

```diff
@@ -585,17 +571,24 @@ cpdef convert_scalar(ndarray arr, object value):
         raise ValueError("cannot set a Timedelta with a non-timedelta "
                          f"{type(value).__name__}")
 
-    if (issubclass(arr.dtype.type, (np.integer, np.floating, np.complex)) and
-            not issubclass(arr.dtype.type, np.bool_)):
+    else:
+        validate_numeric_casting(arr.dtype, value)
+
+    return value
+
+
+cpdef validate_numeric_casting(dtype, object value):
+    # Note: we can't type dtype as cnp.dtype because that cases dtype.type
+    #  to integer
+    if (issubclass(dtype.type, (np.integer, np.floating, np.complex)) and
+            not issubclass(dtype.type, np.bool_)):
         if util.is_bool_object(value):
             raise ValueError("Cannot assign bool to float/integer series")
 
-    if issubclass(arr.dtype.type, (np.integer, np.bool_)):
+    if issubclass(dtype.type, (np.integer, np.bool_)):
```

Comment:
i think much more clear if you reverse these no? (L588 first), then bool casting is out of the way.

Reply:
I'm not clear on what you're suggesting here.  There is no bool casting in this function.

Reply:
reverse the order of the if elif, IOW handle the bool dtype.type first